File: src/sessions.ts
import type { AgentInputItem, Session } from "@openai/agents";
import type { Database, SqlExecutor } from "./database.js";
import { ConflictError, NotFoundError, required } from "./database.js";
import { checksum } from "./crypto.js";
import { hashJson, newId } from "./serialization.js";

export const projectionChecksum = (segments: unknown): string =>
  hashJson(segments);
export const isProjectionChecksumValid = (
  segments: unknown,
  expected: string,
): boolean =>
  expected === projectionChecksum(segments) ||
  expected === checksum(JSON.stringify(segments));

export interface SessionRow extends Record<string, unknown> {
  id: string;
  tenant_id: string;
  status: string;
  scope: Record<string, unknown>;
  last_item_seq: number;
  revision: number;
  active_projection_revision: number;
  created_at: string;
  updated_at: string;
}
export interface SessionItemRow extends Record<string, unknown> {
  id: string;
  session_id: string;
  seq: number;
  item_json: AgentInputItem;
  active: boolean;
  created_at: string;
}

export class SessionService {
  constructor(private readonly db: Database) {}
  async create(tenantId: string, scope: Record<string, unknown> = {}) {
    return required<SessionRow>(
      this.db,
      "INSERT INTO sessions(id,tenant_id,scope) VALUES($1,$2,$3::jsonb) RETURNING *",
      [newId(), tenantId, JSON.stringify(scope)],
    );
  }
  async get(tenantId: string, id: string) {
    return required<SessionRow>(
      this.db,
      "SELECT * FROM sessions WHERE id=$1 AND tenant_id=$2",
      [id, tenantId],
      "session not found",
    );
  }
  async list(tenantId: string, limit = 100, offset = 0) {
    return (
      await this.db.query<SessionRow>(
        "SELECT * FROM sessions WHERE tenant_id=$1 ORDER BY updated_at DESC LIMIT $2 OFFSET $3",
        [tenantId, Math.min(limit, 500), offset],
      )
    ).rows;
  }
  async rawItems(
    sessionId: string,
    options: { activeOnly?: boolean; limit?: number; afterSeq?: number } = {},
  ) {
    const params: unknown[] = [sessionId];
    const conditions = ["session_id=$1"];
    if (options.activeOnly) {
      conditions.push("active=TRUE");
    }
    if (options.afterSeq) {
      params.push(options.afterSeq);
      conditions.push(`seq>$${params.length}`);
    }
    let limit = "";
    if (options.limit) {
      params.push(options.limit);
      limit = ` LIMIT $${params.length}`;
    }
    return (
      await this.db.query<SessionItemRow>(
        `SELECT * FROM session_items WHERE ${conditions.join(" AND ")} ORDER BY seq${limit}`,
        params,
      )
    ).rows;
  }
  async effectiveItems(sessionId: string): Promise<AgentInputItem[]> {
    const session = await required<SessionRow>(
      this.db,
      "SELECT * FROM sessions WHERE id=$1",
      [sessionId],
      "session not found",
    );
    if (session.active_projection_revision > 0) {
      const projection = await this.db.query<{
        segments_json: Array<{ item_json: AgentInputItem }>;
        source_to_seq: number;
        checksum: string;
      }>(
        "SELECT segments_json,source_to_seq,checksum FROM context_projections WHERE session_id=$1 AND revision=$2 AND status='active'",
        [sessionId, session.active_projection_revision],
      );
      if (
        projection.rows[0] &&
        isProjectionChecksumValid(
          projection.rows[0].segments_json,
          projection.rows[0].checksum,
        )
      ) {
        const summary = projection.rows[0].segments_json.map(
          (segment) => segment.item_json,
        );
        const tail = (
          await this.rawItems(sessionId, {
            afterSeq: projection.rows[0].source_to_seq,
          })
        ).map((row) => row.item_json);
        return [...summary, ...tail];
      }
    }
    return (await this.rawItems(sessionId, { activeOnly: true })).map(
      (row) => row.item_json,
    );
  }
  async append(
    sessionId: string,
    items: AgentInputItem[],
    executor: SqlExecutor = this.db,
  ) {
    if (!items.length) return;
    const session = await required<SessionRow>(
      executor,
      "SELECT * FROM sessions WHERE id=$1 FOR UPDATE",
      [sessionId],
      "session not found",
    );
    let seq = session.last_item_seq;
    for (const item of items) {
      seq += 1;
      await executor.query(
        "INSERT INTO session_items(id,session_id,seq,item_json) VALUES($1,$2,$3,$4::jsonb)",
        [newId(), sessionId, seq, JSON.stringify(item)],
      );
    }
    await executor.query(
      "UPDATE sessions SET last_item_seq=$2,revision=revision+1,updated_at=now() WHERE id=$1",
      [sessionId, seq],
    );
  }
  async appendTransactional(sessionId: string, items: AgentInputItem[]) {
    await this.db.transaction((tx) => this.append(sessionId, items, tx));
  }
  async pop(sessionId: string): Promise<AgentInputItem | undefined> {
    return this.db.transaction(async (tx) => {
      const row = (
        await tx.query<SessionItemRow>(
          "SELECT * FROM session_items WHERE session_id=$1 ORDER BY seq DESC LIMIT 1 FOR UPDATE",
          [sessionId],
        )
      ).rows[0];
      if (!row) return undefined;
      await tx.query("DELETE FROM session_items WHERE id=$1", [row.id]);
      await tx.query(
        "UPDATE context_projections SET status='invalidated',updated_at=now() WHERE session_id=$1",
        [sessionId],
      );
      await tx.query(
        "UPDATE compactions SET status='invalidated',failure_reason='session history was destructively modified',updated_at=now() WHERE session_id=$1",
        [sessionId],
      );
      await tx.query(
        "UPDATE sessions SET last_item_seq=COALESCE((SELECT max(seq) FROM session_items WHERE session_id=$1),0),revision=revision+1,active_projection_revision=0,updated_at=now() WHERE id=$1",
        [sessionId],
      );
      return row.item_json;
    });
  }
  async clear(sessionId: string) {
    await this.db.transaction(async (tx) => {
      const result = await tx.query(
        "DELETE FROM session_items WHERE session_id=$1",
        [sessionId],
      );
      await tx.query(
        "UPDATE context_projections SET status='invalidated',updated_at=now() WHERE session_id=$1",
        [sessionId],
      );
      await tx.query(
        "UPDATE compactions SET status='invalidated',failure_reason='session history was cleared',updated_at=now() WHERE session_id=$1",
        [sessionId],
      );
      await tx.query(
        "UPDATE sessions SET last_item_seq=0,revision=revision+1,active_projection_revision=0,updated_at=now() WHERE id=$1",
        [sessionId],
      );
      return result;
    });
  }
  async chatMessages(sessionId: string) {
    const rows = await this.rawItems(sessionId);
    return rows.flatMap((row) => {
      const item = row.item_json as unknown as Record<string, unknown>;
      const role = String(
        item.role ?? (item.type === "message" ? "assistant" : "unknown"),
      );
      const content = item.content;
      let text = "";
      if (typeof content === "string") text = content;
      else if (Array.isArray(content))
        text = content
          .map((part) =>
            typeof part === "string"
              ? part
              : String((part as Record<string, unknown>).text ?? ""),
          )
          .join("");
      if (!text && typeof item.text === "string") text = item.text;
      return text
        ? [
            {
              id: row.id,
              seq: row.seq,
              role,
              content: text,
              item: item,
              created_at: row.created_at,
            },
          ]
        : [];
    });
  }
}

export class DatabaseSession implements Session {
  constructor(
    private readonly service: SessionService,
    private readonly id: string,
  ) {}
  async getSessionId() {
    return this.id;
  }
  async getItems(limit?: number) {
    const items = await this.service.effectiveItems(this.id);
    return limit === undefined ? items : items.slice(-limit);
  }
  async addItems(items: AgentInputItem[]) {
    await this.service.appendTransactional(this.id, items);
  }
  async popItem() {
    return this.service.pop(this.id);
  }
  async clearSession() {
    await this.service.clear(this.id);
  }
}

export function sessionItemText(item: unknown): string {
  if (typeof item === "string") return item;
  if (!item || typeof item !== "object") return String(item ?? "");
  const record = item as Record<string, unknown>;
  if (typeof record.text === "string") return record.text;
  if (typeof record.content === "string") return record.content;
  if (Array.isArray(record.content))
    return record.content.map(sessionItemText).filter(Boolean).join("\n");
  return "";
}

File: src/agent_system/sessions.py
from __future__ import annotations

import json
from typing import Any

from sqlalchemy import select

from .db import Database
from .models import (
    ContextProjectionRecord,
    ContextProjectionSegmentRecord,
    SessionItemRecord,
    SessionRecord,
)
from .serialization import to_jsonable


def session_item_text(value: Any) -> str:
    """Extract display text from an Agents SDK message or content item."""
    if value is None:
        return ""
    if isinstance(value, str):
        return value
    if isinstance(value, list):
        return "".join(session_item_text(item) for item in value)
    if not isinstance(value, dict):
        return str(value)
    if isinstance(value.get("text"), str):
        return str(value["text"])
    if isinstance(value.get("content"), (str, list, dict)):
        return session_item_text(value["content"])
    if isinstance(value.get("output"), (str, list, dict)):
        return session_item_text(value["output"])
    return ""


class DatabaseSession:
    """Agents SDK Session backed by an immutable transcript and active projection.

    ``session_items`` is the canonical transcript. Compaction never clears,
    rewrites, or inserts a fake summary into that table. ``get_items`` overlays
    the latest active context projection with transcript rows appended after the
    projection snapshot.

    The class deliberately does *not* implement ``run_compaction``. Defining it
    would make the Agents SDK treat this store as an
    ``OpenAIResponsesCompactionAwareSession`` and invoke a platform-specific
    protocol that this repository cannot implement atomically.
    """

    session_settings = None

    def __init__(self, db: Database, session_id: str):
        self.db = db
        self.session_id = session_id

    async def clone(self, *, scope: dict[str, Any] | None = None) -> SessionRecord:
        """Atomically clone the active canonical transcript into a new session.

        Context projections are intentionally not copied: a clone starts from the
        same durable source transcript and can then be compacted independently.
        """
        async with self.db.sessions() as session, session.begin():
            source = await session.scalar(
                select(SessionRecord)
                .where(SessionRecord.id == self.session_id)
                .with_for_update()
            )
            if source is None:
                raise KeyError("session not found")
            rows = (
                await session.scalars(
                    select(SessionItemRecord)
                    .where(
                        SessionItemRecord.session_id == self.session_id,
                        SessionItemRecord.active.is_(True),
                    )
                    .order_by(SessionItemRecord.seq.asc())
                )
            ).all()
            cloned_scope = dict(source.scope or {})
            cloned_scope.update(scope or {})
            cloned_scope["cloned_from_session_id"] = self.session_id
            clone = SessionRecord(
                tenant_id=source.tenant_id,
                status="active",
                scope=cloned_scope,
                last_item_seq=max((row.seq for row in rows), default=0),
                revision=1 if rows else 0,
                active_projection_revision=0,
            )
            session.add(clone)
            await session.flush()
            for row in rows:
                session.add(
                    SessionItemRecord(
                        session_id=clone.id,
                        seq=row.seq,
                        item_json=to_jsonable(row.item_json),
                        active=True,
                    )
                )
        return clone

    async def _active_projection(self, session) -> ContextProjectionRecord | None:
        return await session.scalar(
            select(ContextProjectionRecord)
            .where(
                ContextProjectionRecord.session_id == self.session_id,
                ContextProjectionRecord.status == "active",
            )
            .order_by(ContextProjectionRecord.revision.desc())
            .limit(1)
        )

    async def get_items(self, limit: int | None = None) -> list[dict]:
        async with self.db.sessions() as session:
            projection = await self._active_projection(session)
            if projection is None:
                rows = (
                    await session.scalars(
                        select(SessionItemRecord)
                        .where(
                            SessionItemRecord.session_id == self.session_id,
                            SessionItemRecord.active.is_(True),
                        )
                        .order_by(SessionItemRecord.seq.asc())
                    )
                ).all()
                items = [row.item_json for row in rows]
                protected_prefix_count = 0
            else:
                segments = (
                    await session.scalars(
                        select(ContextProjectionSegmentRecord)
                        .where(ContextProjectionSegmentRecord.projection_id == projection.id)
                        .order_by(ContextProjectionSegmentRecord.position.asc())
                    )
                ).all()
                tail = (
                    await session.scalars(
                        select(SessionItemRecord)
                        .where(
                            SessionItemRecord.session_id == self.session_id,
                            SessionItemRecord.active.is_(True),
                            SessionItemRecord.seq > projection.source_to_seq,
                        )
                        .order_by(SessionItemRecord.seq.asc())
                    )
                ).all()
                items = [segment.item_json for segment in segments] + [
                    row.item_json for row in tail
                ]
                protected_prefix_count = 0
                for segment in segments:
                    if segment.segment_type not in {
                        "control_ref",
                        "native_checkpoint",
                        "portable_checkpoint",
                    }:
                        break
                    protected_prefix_count += 1

        if limit is not None and len(items) > limit:
            if protected_prefix_count:
                protected = items[:protected_prefix_count]
                ordinary_limit = max(limit - len(protected), 0)
                items = protected + (items[-ordinary_limit:] if ordinary_limit else [])
            else:
                items = items[-limit:]
        return items

    async def add_items(self, items: list[Any]) -> None:
        if not items:
            return
        async with self.db.sessions() as session, session.begin():
            record = await session.scalar(
                select(SessionRecord).where(SessionRecord.id == self.session_id).with_for_update()
            )
            if record is None:
                raise KeyError("session not found")
            for item in items:
                record.last_item_seq += 1
                session.add(
                    SessionItemRecord(
                        session_id=self.session_id,
                        seq=record.last_item_seq,
                        item_json=to_jsonable(item),
                    )
                )
            record.revision = int(record.revision or 0) + 1

    async def pop_item(self) -> dict | None:
        async with self.db.sessions() as session, session.begin():
            projection = await self._active_projection(session)
            minimum_seq = projection.source_to_seq if projection else 0
            row = await session.scalar(
                select(SessionItemRecord)
                .where(
                    SessionItemRecord.session_id == self.session_id,
                    SessionItemRecord.active.is_(True),
                    SessionItemRecord.seq > minimum_seq,
                )
                .order_by(SessionItemRecord.seq.desc())
                .limit(1)
                .with_for_update()
            )
            if row is None:
                return None
            row.active = False
            record = await session.get(SessionRecord, self.session_id)
            if record is not None:
                record.revision = int(record.revision or 0) + 1
            return row.item_json

    async def clear_session(self) -> None:
        async with self.db.sessions() as session, session.begin():
            record = await session.scalar(
                select(SessionRecord).where(SessionRecord.id == self.session_id).with_for_update()
            )
            if record is None:
                raise KeyError("session not found")
            rows = (
                await session.scalars(
                    select(SessionItemRecord).where(
                        SessionItemRecord.session_id == self.session_id,
                        SessionItemRecord.active.is_(True),
                    )
                )
            ).all()
            for row in rows:
                row.active = False
            projections = (
                await session.scalars(
                    select(ContextProjectionRecord).where(
                        ContextProjectionRecord.session_id == self.session_id,
                        ContextProjectionRecord.status == "active",
                    )
                )
            ).all()
            for projection in projections:
                projection.status = "superseded"
            record.revision = int(record.revision or 0) + 1
            record.active_projection_revision = 0

    async def canonical_items(self, *, include_inactive: bool = False) -> list[dict]:
        """Return the source transcript for audit/recovery, never the projection."""
        async with self.db.sessions() as session:
            query = select(SessionItemRecord).where(SessionItemRecord.session_id == self.session_id)
            if not include_inactive:
                query = query.where(SessionItemRecord.active.is_(True))
            rows = (await session.scalars(query.order_by(SessionItemRecord.seq))).all()
        return [
            {"id": row.id, "seq": row.seq, "item": row.item_json, "active": row.active}
            for row in rows
        ]

    async def chat_messages(self) -> list[dict[str, Any]]:
        """Return canonical user/assistant messages with reasoning kept separate."""
        canonical = await self.canonical_items()
        messages: list[dict[str, Any]] = []
        pending_reasoning: list[str] = []
        pending_reasoning_kind = "raw"
        for row in canonical:
            item = dict(row["item"] or {})
            item_type = str(item.get("type") or "")
            role = str(item.get("role") or "")
            if item_type == "reasoning":
                summary = session_item_text(item.get("summary"))
                raw = session_item_text(item.get("content"))
                if summary:
                    pending_reasoning.append(summary)
                    pending_reasoning_kind = "summary"
                elif raw:
                    pending_reasoning.append(raw)
                    pending_reasoning_kind = "raw"
                continue
            if role not in {"user", "assistant"}:
                continue
            content = session_item_text(item.get("content"))
            if not content and role == "assistant":
                content = session_item_text(item.get("output"))
            message = {
                "id": str(item.get("id") or row["id"]),
                "seq": row["seq"],
                "role": role,
                "content": content,
                "status": str(item.get("status") or "completed"),
            }
            if role == "assistant" and pending_reasoning:
                message["reasoning"] = "\n\n".join(pending_reasoning)
                message["reasoning_kind"] = pending_reasoning_kind
                pending_reasoning = []
                pending_reasoning_kind = "raw"
            messages.append(message)
        return messages


def estimate_items_size(items: list[dict]) -> int:
    """Legacy display-only byte size; never used for compaction decisions."""
    return sum(len(json.dumps(item, ensure_ascii=False).encode("utf-8")) for item in items)
